tajo-core/src/main/java/org/apache/tajo/querymaster/QueryMaster.java

```diff
@@ -286,6 +292,47 @@ public void stopQuery(QueryId queryId) {
 
       finishedQueryMasterTasksCache.put(queryId, queryMasterTask);
 
+      // In INSERT .. SELECT ... queries or CTAS, dynamic partitions should be added to catalog.
+      Query query = queryMasterTask.getQuery();
+      if (query.getState() == TajoProtos.QueryState.QUERY_SUCCEEDED && query.getResultDesc() != null) {
```

Comment:
Here is not good place to do this because we cannot guarantee the query atomicity. In other words, a query is successfully completed but the partition additions can be failed. Basically, we have to guarantee both the query completion and partition addition or nothing.

QueryCompletedTransition::finalizeQuery in Query is the best place for this.